Zend/zend_objects.c

```diff
@@ -303,12 +339,33 @@ ZEND_API void ZEND_FASTCALL zend_objects_clone_members(zend_object *new_object,
 	}
 }
 
-ZEND_API zend_object *zend_objects_clone_obj(zend_object *old_object)
+ZEND_API void ZEND_FASTCALL zend_objects_clone_members(zend_object *new_object, zend_object *old_object)
+{
+	ZEND_ASSERT(old_object->ce == new_object->ce);
+
+	zend_objects_clone_members_ex(new_object, old_object, old_object->ce, &zend_empty_array);
+}
+
+ZEND_API zend_object *zend_objects_clone_obj_with(zend_object *old_object, zend_class_entry *scope, const HashTable *properties)
 {
 	zend_object *new_object;
 
+	/* Compatibility with code that only overrides clone_obj. */
+	if (UNEXPECTED(old_object->handlers->clone_obj != zend_objects_clone_obj)) {
```

Comment:
Isn't this problematic for overridden `clone_obj` that reuse `zend_objects_clone_obj`? In that case `old_object->handlers->clone_obj != zend_objects_clone_obj` would hit, and we'd recurse infinitely.

Reply:
Yes. There are none in php-src and generally speaking it seems unlikely that there are any, since when there is a need to overwrite `clone` there likely also is a need to overwrite `create` (which explicitly isn't supported):

> /* assume that create isn't overwritten, so when clone depends on the
> * overwritten one then it must itself be overwritten */

Most of the custom clone handlers just use `zend_objects_clone_members()`.

------------------

This PR tries to minimize the disruption for existing code, but if there is an external extension that does what you suggest, then it's easy enough for the author to adjust it to add an explicit `clone_with` handler.

Reply:
> This is likely dangerous, since some internal classes might expect their (readonly) properties not to change

Can this already be broken with `__clone` in a user-declared sub-class? I can't see cases other than enums preventing the declaration of `__clone`. The property would need to be `private(set)` to prevent this.

Reply:
Well, the new commit looks much more logical to me, and it should work automatically for extensions that already customize clone. Right now I'm failing to see the downsides, but maybe some others can chime in.

Reply:
> Right now I'm failing to see the downsides,

For me it made sense to have the “clone-with” logic as part of the `zend_objects_clone_members` to have all the cloning logic in a single place, particularly for classes that *do* need a custom `clone_with` and would have to reimplement it with the new commit.

Reply:
I agree with Ilija on this. I like the current code more.
You could split off the property update logic to a separate helper so that part can be reused? I'd think that either:
- A custom class may want to do some validation and call such a helper
- A custom class has entire custom logic to deal with it

Reply:
Okay. I've also gotten private feedback that the new version is preferable, so I'll leave it like that. Thanks!